xformers/components/nvfuser/bias_dropout_res_layernorm.py

```diff
@@ -0,0 +1,91 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+
+from typing import Optional
+
+import torch
+import torch.nn as nn
+from functorch.compile import memory_efficient_fusion
+
+from xformers.components import LayerNormStyle
+
+
+def _fn(
+    x: torch.Tensor,
+    bias: Optional[torch.Tensor],
+    prob: float,
+    layer_norm_style: Optional[LayerNormStyle],
+    norm: nn.Module,
+    res: torch.Tensor,
+) -> torch.Tensor:
+    a = torch.add(x, bias) if bias is not None else x
+    b = torch.nn.functional.dropout(a, prob) if prob > 0.0 else a
+    if layer_norm_style == LayerNormStyle.Pre:
+        c = norm(b)
+        return torch.add(c, res)
+    elif layer_norm_style == LayerNormStyle.Post:
+        c = torch.add(b, res)
+        return norm(c)
+    else:
+        raise ValueError
+
+
+class NVFusedBiasDropoutResLayerNorm(torch.nn.Module):
+
+    """
+    A layer which fuses the computation of LayerNorm, Residual, and Dropout(x + Bias)
+    operations with AOTAutograd and nvFuser based on specified layer norm style
+    """
+
+    def __init__(
+        self,
+        p: float,
+        d_model: int,
+        bias_shape: Optional[int] = None,
+        layer_norm_style: LayerNormStyle = LayerNormStyle.Post,
+    ) -> None:
+        super().__init__()
+
+        self.p = float(p)
+        self.requires_residual = True
+        self.layer_norm_style = layer_norm_style
+
+        self.bias = (
+            nn.Parameter(torch.zeros(bias_shape, device=torch.device("cuda")))
+            if bias_shape is not None
+            else None
+        )
+        self.norm = nn.LayerNorm(d_model)
+
+        assert (
+            self.p < 1.0
+        ), f"We don't want to drop all the values, most probably p={self.p} is not properly set"
+
+    def init_weights(self, *args, **kwargs):
+        with torch.no_grad():
+            if self.bias is not None:
+                self.bias.fill_(0.0)
+
+    def forward(self, x: torch.Tensor, res: torch.Tensor) -> torch.Tensor:
+        # Train/inference
+        p = self.p if self.training else 0.0
+
+        # Catch possible type or device mismatch
+        self.norm = self.norm.to(dtype=x.dtype, device=x.device)
+        if self.bias is not None and self.bias.dtype != x.dtype:
+            self.bias = nn.Parameter(
+                torch.zeros_like(self.bias, dtype=x.dtype, device=x.device)
+            )
```

Comment:
This shouldn't be in the `forward`. Parameter initialization should happen in the constructor, and the `norm` cast should be handled naturally given that `norm` is a `nn.Module`